### PR TITLE
ENH: Add Hamilton's FFTLog.

### PR DESCRIPTION
**Pull request by prisae**
Adding a f2py-wrapper for the logarithmic FFT by A. J. S. Hamilton;
original code available at: http://casa.colorado.edu/~ajsh/FFTLog.

---

This is an initial pull request to add Hamilton's FFTLog to the fftpack of SciPy, as discussed on the mailing list starting with https://mail.scipy.org/pipermail/scipy-dev/2016-October/021552.html.

Some important points:
- **fftpack/src/fftlog/cdgamma.f**: It has to be replaced with `scipy.special.loggamma` (I am lacking the necessary knowledge to do this) => all calls `cdgamma(x, 1)` in `fftlog.f` can be replaced with `loggamma(x)`. See discussion here and in the whole thread in general: https://mail.scipy.org/pipermail/scipy-dev/2016-October/021584.html.

- **fftpack/FFTLog-Example.ipyn**: Not part of it, they are just examples of its usage.
 
I tried to follow the *Contributing to SciPy* (HACKING.rst.txt) guide as much as possible. In terms of the provided checklist in HACKING.rst.txt, I did all but:

- Mention new functionality in release notes of next release
- Adding new functionality to reference guide

I welcome any feedback!
Dieter

### CODE REVIEW

**Comment by prisae** [comment]
Link to scipy-dev message: https://mail.python.org/pipermail/scipy-dev/2017-April/021939.html

**Comment by rgommers** [comment]
> Mention new functionality in release notes of next release

Don't worry about this for now, easier to do right before or after merging.

> Adding new functionality to reference guide

The continuous integration will complain about this. I'll have a look at what to add where.

**Comment by rgommers** [comment]
There's a couple of doctest failures, and a non-ascii character. Have a look at the bottom of https://travis-ci.org/scipy/scipy/jobs/222840699 for details

File: scipy/fftpack/logarithmic.py [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+from .basic import _asfarray
+
+__all__ = ['fftlog', 'fftlogargs']
+
+atexit.register(_fftl.destroy_fftl_x_cache)
+atexit.register(_fftl.destroy_fftl_w_cache)
+del atexit
+
+
+def fftlog(x, dlogr, mu='sine', q=0.0, kr=1.0, rk=1.0):
```
Comment (rgommers) [inline scipy/fftpack/logarithmic.py]: Are all these parameters really necessary, and if so can they be given more descriptive names? Especially ``kr`` and ``rk`` seem like parameters that are taken over from the Fortran code but are unintuitive for users.

File: scipy/fftpack/logarithmic.py [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+        output array by sqrt(rk)^dir, so if you want to do the normalization
+        later, or you don't care about the normalization, you can set rk = 1.
+        Default is 1.
+
+    n : int, optional
+        Defines the length of the Fourier transform.  If `n` is not specified
+        (the default) then ``n = x.shape[axis]``.  If ``n < x.shape[axis]``,
+        `x` is truncated, if ``n > x.shape[axis]``, `x` is zero-padded.
+
+    axis : int, optional
```
Comment (rgommers) [inline scipy/fftpack/logarithmic.py]: this parameter as well as ``overwrite_x`` is not in the signature of this function

File: scipy/fftpack/logarithmic.py [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+        last axis.
+
+    overwrite_x : bool, optional
+        If set to true, the contents of `x` can be overwritten. Default is
+        False.
+
+    Returns
+    -------
+    y : real ndarray
+        Transformed array Ã(k): a(j) is Ã(k_j) at k_j = k_c exp[(j-jc) dlnr].
```
Comment (rgommers) [inline scipy/fftpack/logarithmic.py]: These weird A's are non-ascii and should be removed.

File: scipy/fftpack/logarithmic.py [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+    ----------
+    .. [1] 'Uncorrelated modes of the non-linear power spectrum', by A. J. S.
+           Hamilton, `Monthly Notices of the Royal Astronomical Society` vol.
+           312, pp. 257-284, http://dx.doy.org/10.1046/j.1365-8711.2000.03071.x
+           (2000). Website of FFTLog: http://casa.colorado.edu/~ajsh/FFTLog.
+
+    Examples
+    --------
+    >>> from scipy.fftpack import fftlog, fftlogargs
+    >>> # Get fftlog-arguments
```
Comment (rgommers) [inline scipy/fftpack/logarithmic.py]: Instead of comments you should use normal text (i.e. remove ``>>> # ``) with blank lines above and below.

Comment (rgommers) [inline scipy/fftpack/logarithmic.py]: that goes for all comments in here

File: scipy/fftpack/logarithmic.py [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+    >>> fw = np.sqrt(np.pi/2/w)  # Frequency domain
+    >>> ft = 1/np.sqrt(t)        # Time domain
+    >>> # FFTLog
+    >>> fftl = fftlog(fw, dlogr=dlogr, mu=mu, q=q, kr=kr, rk=rk)
+    >>> fftl *= 2/np.pi  # Scale back
+    >>> # Print result
+    >>> print('Input      :', fw)
+    >>> print('Analytical :', ft)
+    >>> print('fftlog     :', fftl)
+    Input      : [ 1.48956664  1.32757767  1.18320484  1.05453243]
```
Comment (rgommers) [inline scipy/fftpack/logarithmic.py]: the output should be straight under each print statement

File: scipy/fftpack/logarithmic.py [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+    if np.iscomplexobj(tmp):  # Returns complex128
+        y = (_fftl.drfftl(tmp.real, n, nmu, q, dlnr, kr, rk, 1) +
+             1j*_fftl.drfftl(tmp.imag, n, nmu, q, dlnr, kr, rk, 1))
+    else:  # Returns float64
+        y = _fftl.drfftl(tmp, n, nmu, q, dlnr, kr, rk, 1)
+
+    return y
+
+
+def fftlogargs(n, dlogr=0.01, logrc=0.0, mu='sine', q=0, kr=1, kropt=0):
```
Comment (rgommers) [inline scipy/fftpack/logarithmic.py]: Related to my API question for ``fftlog``: why is this function needed at all? Can you instead use this internally in ``fftlog``?

File: scipy/fftpack/logarithmic.py [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+    >>> from scipy import fftpack
+    >>> n = 4
+    >>> kr = 1
+    >>> kropt = 1
+    >>> outpts = fftpack.fftlogargs(n, kr=kr, kropt=kropt)
+    >>> print('intpts :', outpts[0])
+    >>> print('outpts :', outpts[1])
+    >>> print('kr     :', outpts[2])
+    >>> print('rk     :', outpts[3])
+    intpts : [ 0.96605088  0.98855309  1.01157945  1.03514217]
```
Comment (rgommers) [inline scipy/fftpack/logarithmic.py]: same as for ``fftlog``, put print outputs straight under each print statement

**Comment by prisae** [comment]
Thanks for reviewing it Ralf. Few things:

- Doctest-problems are hopefully resolved.
- I thought non-ascii problem is resolved too, but apparently not. I will have another look.
- Yes, all the variable-names were straight from the fortran code. I renamed the following variables:
   - dlogr -> spacing
   - mu -> transform
   - q -> bias
   - logrc -> center

`kr` and `rk` are somewhat difficult to rename. They are the product of the center points of k and r, and r and k respectively (taking into account where `center` is located). Now this leads to the function `fftlogargs`: FFTLog has the possibility to find the best, low-ringing values for `kr` and `rk`.

`fftlogargs` could be included into `fftlog`, and I like the idea. However, `fftlog` would then not only have to return the transformed values, but also the array (abscissae) where these values are. Because if FFTLog calculates the low-ringing values for `kr` and `rk`, but you don't know these values, there is no way for you to know the position of the returned values.

Changes:
- `fftlog(x, spacing, transform, bias, kr, rk)` => `fftlog(x, spacing, transform, bias, center, kropt, kr, rk)`
- Output would be the transformed array and additionally:
   - output-pts (abscissae) of transformed array
   - kr, rk  (optional, but you might want to use the same values again)

`fftlogargs` could then be reduced to an optional helper function:

        return 10**(center + (np.arange(n)+1 - (n + 1)/2.0)*spacing)

to only calculated the required input points for given `center`, `spacing`, and points `n`.

Do you prefer to keep it as it is or shall I make the changes to include `fftlogargs` into `fftlog`?

**Comment by prisae** [comment]
Turns out that the **ü** in my surname was the ascii-problem...

**Comment by prisae** [comment]
Glad to see it added to the [1.1.0](https://github.com/scipy/scipy/milestone/34) milestone. I assume it still needs work, but I am happy to iterate over it.

**Comment by rgommers** [comment]
> I assume it still needs work

Not sure right now. We're short on reviewer power unfortunately, and fftpack isn't the most active submodule. I'm adding PRs with enhancements that seem ready to the milestone for the next release so we make sure to go over those and try to get them merged.

**Comment by tylerjereddy** [comment]
Bumping again -- I'm not the right person to review FFT Fortran code at this time, but hopefully we will regain momentum as Ralf suggests.

**Comment by mreineck** [comment]
I'm sorry ... I've inadvertently sabotaged this PR by pushing for the switch to `pocketfft ` :( There is now no longer any Fortran FFT code in `scipy`, and so integrating this addition will probably be much harder.

I don't have a feeling how hard it would be to implement FFTLog on top of the new C++ module, but I volunteer to assist, if I can.

**Comment by prisae** [comment]
Thanks @mreineck for picking this up. It all depends. Mainly,
- if there is enough interest to have FFTLog in SciPy, and
- if there is an experienced FFT-person willing to review my modest go at it...

In the meantime since I made this PR I created a pure Python-Version of it, which could be implemented instead. It can be found here: https://github.com/prisae/pyfftlog

I never benchmarked FFTLog versus pyfftlog. But there are no loops, and most work is carried out within `scipy.fftpack._fftpack.drfft` and `scipy.special.loggamma`, so I would not expect massive differences.

**Comment by prisae** [comment]
I just realized that, with the new changes to `pocketfft`, `scipy.fftpack._fftpack.drfft` might not be available any longer. That would be a very bad thing for many of my codes. When will this change take effect? Is there an alternative for `scipy.fftpack._fftpack.drfft` which behaves similar?

**Comment by mreineck** [comment]
> I just realized that, with the new changes to pocketfft, scipy.fftpack._fftpack.drfft might not be available any longer.

It isn't, but I think that can be worked around easily.

If I read the code correctly, you are using this function simply to perform a standard FFTPACK-style forward real FFT, correct?

If so, is there any reason why you can't use the higher level `scipy.fftpack.rfft` function? That is still there and its behavior has not changed.

**Comment by prisae** [comment]
That is a good question.

Looking through my commit messages I see that I tried `rfft`/`irfft`, `basic._raw_fft`, and `_fftpack.drfft`, and settled with the last.

According to a message: "replace i/rfft with _fftpack.drfft, ~1/4 speed-up".

So I guess yes, I can replace them with `rfft` and `irfft`.

**Comment by prisae** [comment]
Thanks for the hint @mreineck . I just checked [empymod](https://github.com/empymod/empymod), which is the main package where I use FFTLog. And there I actually still have it implemented as
```
a = fftpack.rfft(a)
a = fftpack.irfft(a)
```
so all good. I just have to adjust it in https://github.com/prisae/pyfftlog/

**Comment by mreineck** [comment]
> According to a message: "replace i/rfft with _fftpack.drfft, ~1/4 speed-up".

Ah, so you were most likely bitten by calling overhead ... with any luck that should not be such a big issue any more.

Great! So in principle your Python code should run fine on top of the new FFTs, with this minimal name change.

If this turns out to be too slow, I could try to port the Python file to C++, but I'd suggest to wait for benchmark numbers first!

**Comment by prisae** [comment]
If you are willing to look at (a new) PR for pyfftlog I will do some benchmarks between FFTLog and pyfftlog, to see whether porting is necessary or not.

- If there is no big difference I would make some changes to pyfftlog (basically include the feedback I got in the review of the current PR by @rgommers), and create a new PR for pyfftlog.
- If there is a big difference then it would be great if you could try to port it to C++.

But again, I will only push further on this if you are committed to look at a potential PR. So I don't plan to invest any further time unless someone is interested in it...

**Comment by mreineck** [comment]
While I'd certainly be happy to get involved in the PR, I'm not a `scipy` member, so I don't have any influence on whether it will be accepted...

**Comment by prisae** [comment]
That is good enough for me. It was in general given the green light in the mailing list (green light in the sense that it could be of interest to have it in SciPy), and Ralf reviewed the code once. The reason it was not included was that there was not really an active maintainer for fftpack. But now with all the movement with regards to pocketfft this might have changed.

It won't be immediately anyway from my side, I hope I have the time for it within a month or so. I'll ping you.

Also, I keep this PR open until a make a new one with pyfftlog.

**Comment by mreineck** [comment]
> The reason it was not included was that there was not really an active maintainer for fftpack. But now with all the movement with regards to pocketfft this might have changed.

That would indeed be great. @peterbell10 has been doing miracles for FFT support in `scipy` as his GSoC project; I hope he will continue having an eye on this code after GSoC has ended.

**Comment by rgommers** [comment]
> I'm not a `scipy` member, so I don't have any influence on whether it will be accepted...

@mreineck if you'd be interested in changing that, happy to chat:) either way, your expertise is much appreciated around here, so you do have influence!

**Comment by mreineck** [comment]
> @mreineck if you'd be interested in changing that, happy to chat:) either way, your expertise is much appreciated around here, so you do have influence!

That's very kind, thanks! Overall I think the current arrangement is perfectly fine; I just didn't want to give the false impression that I have any official role in the decision-making process :)

**Comment by prisae** [comment]
### Benchmark

I run the promised benchmark, @mreineck :

benchmark

Looks like it would indeed profit from a C++ implementation. I usually use `n` in the range of 20-50, so rather low.

However, two comments:
- This benchmark was carried out with `scipy 1.13.0`; the benchmark might change with the new pocketfft - could you run the benchmark on your machine to check?
- We are in the microseconds. So the calculations around the FFTLog are most likely slower than the FFTLog itself, and it might not be worth optimising it.

What are your thoughts? Do you think we should go with the Python FFTLog, or with a C++ implementation of it?

You can find the benchmark here:  
https://github.com/prisae/tmp-share/blob/master/benchmark_fftlog_pyfftlog.ipynb

**Comment by prisae** [comment]
Also, the benchmark shows that they are approaching the same line -- for really large arrays they will become similar, as the same fft-routines are at the core of the algorithm. So the slowness of fftlog must really come from some overhead. But then, the whole purpose of FFTLog is that you do not have to use too large arrays. In my use case 5-10 points per decade are sufficient, and only a few decades are required. However, I am sure there will be other use cases where this does not apply.

**Comment by larsoner** [comment]
Let's avoid C++ unless we really need it. It increases maintenance burden. (It made sense for core `fft` functions since these are used everywhere.) It might be better to first optimize the Python code itself first in Python. It still uses the old fftpack interface `scipy.fftpack.rfft` and real/imag conventions, which could be slower than using `scipy.fft.rfft` instead. If that still proves too slow, Cython is more easily maintained than C++ as well.

Here is what I get for speed:

benchmark

**Comment by prisae** [comment]
Yes, I agree, and that is fine with me. There seems to be an overhead of 40 us in your example, which is OK I think.

I will try to flesh out a new PR with the pure Python version. 2.5 years passed since the original PR, so I should be able to make a better/cleaner PR now than back then.

**Comment by mreineck** [comment]
> It still uses the old fftpack interface scipy.fftpack.rfft and real/imag conventions, which could be slower than using scipy.fft.rfft instead.

They are actually faster :) `pypocketfft` uses exactly this format internally and provides an interface which presents the data in this format, and this is used by `scipy.fftpack` AFAIK. So no need to change anything here.

**Comment by peterbell10** [comment]
An observation: The original fortran routines only operate on 1d arrays and all of the code to adapt these into multi-dimensional operations is currently on the C++ side. Similarly, this is where vectorization and multithreading are handled. If FFTLog is done in C++ then these extra features should come almost for free.

@prisae does pyfftlog work with multidimensional arrays? It looks like a more or less straight translation of the 1d fortran version but I suppose numpy broadcasting rules might just make it work.

**Comment by prisae** [comment]
@peterbell10 - I don't know, and my understanding of FFT algorithms is not sufficiently sound to make an appropriate guess. FFTLog works fine for my needs, which is purely 1D. This is also why it would be good if someone with a deeper understanding of FFT algorithms would have a look at the implementation.

I do think the FFTLog is an amazing algorithm, and could be used wider than it currently is. This is why I thought it might be very nice to have it in SciPy.

**Comment by mreineck** [comment]
> @prisae  does pyfftlog work with multidimensional arrays? It looks like a more  or less straight translation of the 1d fortran version but I suppose  numpy broadcasting rules might just make it work.

Frankly, I would not generalize this without a concrete use case. Especially since `r2r` transforms tend to behave in unexpected ways when going from 1D to multi-d.

**Comment by peterbell10** [comment]
> I would not generalize this without a concrete use case. Especially since r2r transforms tend to behave in unexpected ways when going from 1D to multi-d.

Note that there's a distinction between "works with mutidimensional arrays" and a "multidimensional transform". What I expect is that `fftlog(x)` should apply seperate 1d transforms over a single axis within the multidimensional array, in the same way that `fft(x)` would.

**Comment by prisae** [comment]
Sorry for the delay, I am currently missing the bandwidth, and I most likely won't get to this in the near to medium future.

If anyone wants to have a go at it I'd be happy to explain as much as I can though (and as much as I remember...).

**Comment by larsoner** [comment]
@peterbell10 do you want to take over?

In the meantime I'm going to remove the 1.4.0 milestone. We can put it back if someone volunteers to work on this in the next few weeks.

**Comment by peterbell10** [comment]
I don't think I'll have enough time to work on this in the near future.

**Comment by ntessore** [comment]
I would like to revive the effort to get FFTLog into scipy, as it seems that people are reimplementing this constantly.

Instead of using the original Fortran FFTLog implementation, why not use a Python implementation that calls whatever FFT routine in scipy is fastest. [Here is an example](https://gist.github.com/ntessore/3dc14068fa07bead5493cbaecb5d9328) of the main routine using `numpy.fft`. As you can see the whole FFTLog algorithm reduces to a couple of lines of code. As someone anticipated correctly, it threads nicely over multidimensional input arrays.

**Comment by prisae** [comment]
If someone picks it up that would be great @ntessore.

For what its worth, my implementations can now be installed via pip and via conda:
```
pip install pyfftlog
```
or
```
conda install -c conda-forge pyfftlog
```
for the pure Python version, or 
```
pip install fftlog
```
or
```
conda install -c conda-forge fftlog
```
for the Python-wrapped Fortran version.

**Comment by prisae** [comment]
The FFTLog algorithm contains possibilities for a Hankel transform and a Fourier transform. As far as I can see, the gist you posted contains only the Hankel transform, right? The name "FFTLog" is a bit misleading, unfortunately.

**Comment by ntessore** [comment]
The gist I posted only contains the main routine as an example of how simple the implementation would be. Every other routine in the original FFTLog is then essentially a one-liner to provide a more convenient interface for special cases. Note that I also only show the forward transform, since the backward transform should be its own function, in the spirit of `rfft/irfft` etc.

**Comment by prisae** [comment]
I would certainly love to have it in SciPy.

**Comment by mreineck** [comment]
Thanks, @ntessore, for the demo! This certainly looks like a good way of having the functionality in scipy.
Only one minor comment: I suggest using `scipy.fft` instead of `numpy.fft` as the underlying Fourier transform implementation, since it is typically much faster for multidimensional arrays. And since the code will be part of `scipy`, we know that it will always be available.

**Comment by prisae** [comment]
(Just FWIW: in my experience the Fortran-wrapped version is usually faster by a margin, so it would be nice to have that one. However, I'd be happy to have a pure Python version over not having anything at all.)

**Comment by mreineck** [comment]
I don't remember exactly, but my recollection is that the Fortran code for FFTlog was fairly closely linked to FFTPACK, which is no longer available in Scipy. So if there should be a compiled version of FFTLog, it would require substantial rewriting. At least for the moment I'd lean towards a Python implementation, since it really appears to be very easy to do. If a compiled implementation follows at a later time, not much effort would be lost.

**Comment by prisae** [comment]
Yes, I think that was one of the reason why this PR stalled in the end @mreineck - given your insights a pure Python version sounds more feasible indeed.

**Comment by ntessore** [comment]
Using your `fftlog` module, I see that the Fortran implementation is roughly 10x faster than the pure Python version, not too bad for an unoptimised example. And a line profile shows that there might be potential for improvement:

```
% Time  Line Contents
==============================
 ...
          # Hankel transform via real FFT
 6.3      g = rfft(f, axis=-1)
 0.6      g *= u
51.0      g = irfft(g, n, axis=-1)
36.2      g[..., :] = g[..., ::-1]
...

```

**Comment by mreineck** [comment]
Any idea why the `irfft` does so much worse than the `rfft`? That should never happen...

**Comment by mreineck** [comment]
> That should never happen...

Hmm, maybe that depends on the transform length. Which length did you use for this run?

**Comment by ntessore** [comment]
> Any idea why the `irfft` does so much worse than the `rfft`? That should never happen...

Apologies, that was the compiler being flaky due to insufficient time resolution for an input size of `128`. For larger inputs, the picture is different. Here's `n = 1024`:

```
Timer unit: 1e-06 s

Total time: 0.001043 s
File: /Users/mbessnt3/code/fftlognt/fftlognt.py
Function: fhtq at line 23

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
...
    60         1        160.0    160.0     15.3      loggamma(u, out=u)
    61         1        277.0    277.0     26.6      loggamma(v, out=v)
...
    75                                               # Hankel transform via real FFT
    76         1        154.0    154.0     14.8      g = rfft(f, axis=-1)
    77         1         13.0     13.0      1.2      g *= u
    78         1        106.0    106.0     10.2      g = irfft(g, n, axis=-1)
    79         1          9.0      9.0      0.9      g[..., :] = g[..., ::-1]
...
```

And pure Python is only about 1/3 as fast as Fotran. For `n = 8192`:

```
Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
...
    75                                               # Hankel transform via real FFT
    76         1        469.0    469.0      5.9      g = rfft(f, axis=-1)
    77         1         66.0     66.0      0.8      g *= u
    78         1        369.0    369.0      4.6      g = irfft(g, n, axis=-1)
    79         1         30.0     30.0      0.4      g[..., :] = g[..., ::-1]
...
```

Now pure Python is 2/3 as fast as Fortran. So I guess tweaking the `loggamma` calls would make more sense than eventually going back to Fortran.

**Comment by mreineck** [comment]
Interesting, thanks! It almost seems like `loggamma` is having a harder and harder job computing its result with increasing length, since its time consumption increase is superlinear.

**Comment by ntessore** [comment]
I suspect that it could do with a lot less precision here than in the general purpose case. If speed is really that important, perhaps it would make sense to port the `cdgamma` Fortran implementation that FFTLog uses, since some consideration about the necessary precision would have gone into that.

Lastly, I wanted to point out that the Fortran implementation is really not any more complicated than the Python example. Almost all of the complexity comes from setting up the FFT, and in particular dealing with the last element of the array, which the Python version just hands over to `rfft/irfft`.

**Comment by mreineck** [comment]
The complexity (or let's say the maintenance burden) will come mostly from the interface, I think. If I look at the amount of glue code that comes between the pure C++ FFT implementation in `scipy` and the actual user-visible functions in `scipy.fft`, this is comparable in size to the code which actually does the numerical work.

But I should stress that this is only my personal opinion - I'm only involved in `scipy` as a code contributor and don't get to decide how new components should be added :-)

**Comment by larsoner** [comment]
+1 for a pure-Python solution. I say let's get something that works properly with a good/suitable interface first. That is enough work to start. After that, we can have follow-up PRs to make it more efficient where there are pain points. For example the line `g[..., :] = g[..., ::-1]` can probably just be `g = g[..., ::-1]` (unless we need C-contiguity for some reason) and it's essentially free because it's just creating a view. But these sorts of things are better to do after it already works -- it's easier for other people to contribute or even move parts or all of it to Cythton once there's a working API, implementation, and tests all in place.

**Comment by ntessore** [comment]
Please let me know how I can help. I can provide a pure Python implementation of the classical FFTLog functionality which has the following interface:

- `dctl`, `idctl`
  computes the discrete Fourier cosine transform of a logarithmically spaced periodic sequence.
  This is a driver routine that calls `fhtq`.
- `dstl`, `idstl`
  computes the discrete Fourier sine transform of a logarithmically spaced periodic sequence.
  This is a driver routine that calls `fhtq`.
- `fht`, `ifht`
  computes the discrete Hankel transform of a logarithmically spaced periodic sequence.
  This is a driver routine that calls `fhtq`.
- `fhtq`, `ifhtq`
  computes the biased discrete Hankel transform of a logarithmically spaced periodic sequence.
  This is the basic FFTLog routine.

**Comment by larsoner** [comment]
That looks reasonable to me -- @ntessore @prisae thoughts?

@mreineck @peterbell10 I guess these should live in `scipy.fft`?

**Comment by mreineck** [comment]
> @mreineck @peterbell10 I guess these should live in scipy.fft?

Feels like the most natural place, yes.

Thanks a lot, @ntessore, for taking care of this!

**Comment by prisae** [comment]
Sounds OK to me. I think the best is if you simply provide the functionality and then we can iterate over it. Feel free to use this PR and replace all the files, or start a new one. We can then link to this one and close this. Either way is fine. I am happy to review with what you come up.

**Comment by larsoner** [comment]
Given how much of this PR has to do with wrapping to Fortran, I would probably just keep parts of the `scipy/fftpack/tests/test_logarithmic.py` test after moving it to `scipy/fft/tests/test_logarithmic.py`. But @ntessore I'll let you decide.

I'll close this but you should be able to pull commits from @prisae's branch if need be

**Comment by ntessore** [comment]
Ok thanks. I will pull in prisae/include-fftlog and move stuff around, that should credit @prisae as well. The I will open a new PR. Give a thumbs up to this and I will ping you in the new PR.

**Comment by larsoner** [comment]
Yes feel free to do that. In the end I will probably merge your PR with squash+merge and let the "Co-authored by" take care of dual code credit. The alternative is to `git rebase -i HEAD~#` and squash @prisae's exsiting commits, then at the end of your PR you squash all of yours, but this is less than ideal because once those are merged to master the 3 squashed commits from his PR are essentially in a broken state, which is not good if someone has to do a `git bisect` later.

**Comment by prisae** [comment]
Thanks @ntessore , looking forward to it!

**Pull request closed without merging**